# Off-Ball Run Analysis: Quantifying Space Creation in Football

**Author:** Ivo Steinke  
**Competition:** PySport X SkillCorner Analytics Cup

---

## Introduction

Off-ball movement is crucial to modern football tactics, yet traditional analysis focuses primarily on players in possession. This work presents a novel approach to quantify how high-speed off-ball runs create exploitable space for the ball carrier.

**Research Question:** *How much space do coordinated off-ball movements generate during team possession?*

The methodology analyzes 10 A-League matches using SkillCorner's broadcast tracking data, combining velocity-based run detection with Voronoi tessellation to measure spatial impact. This transparency-first approach provides interpretable metrics for tactical analysis without relying on black-box machine learning.

In [5]:
# Imports
from src.data_loader import load_matches_info, load_match_data
from src.space_analysis import analyze_all_matches_normalized
import pandas as pd

# Configuration
MATCH_IDS = ['2017461', '1996435', '1886347', '1899585', '1925299',
             '1953632', '2006229', '2011166', '2013725', '2015213']
VELOCITY_THRESHOLD = 5.0  # m/s (18 km/h)

print(f"Analyzing {len(MATCH_IDS)} A-League matches")
print(f"Velocity threshold: {VELOCITY_THRESHOLD} m/s")

Analyzing 10 A-League matches
Velocity threshold: 5.0 m/s


## Methodology

### Data
10 A-League matches (2024/25 season) containing 688,000+ player position frames at 10 Hz, focusing on runs during own-team possession phases.

### Run Detection
High-speed off-ball movements identified using three criteria:
- **Velocity threshold:** ≥5.0 m/s (18 km/h)
- **Minimum duration:** 3 seconds sustained movement
- **Context filter:** Own team in possession, runner without ball

### Space Measurement
For each detected run, Voronoi diagrams calculated at run start (t₀) and end (t₀+3s) to quantify controlled area changes. Space creation measured specifically for the ball carrier:
```
Space_Created = Voronoi_Area(ball_carrier, t₀+3s) - Voronoi_Area(ball_carrier, t₀)
```

### Normalization
All runs normalized to attack direction (left-to-right) accounting for teams switching sides at halftime, enabling cross-match spatial pattern analysis.

In [6]:
# Load and analyze all matches
matches = load_matches_info(MATCH_IDS)
trajectories = analyze_all_matches_normalized(matches, load_match_data, 
                                              velocity_threshold=VELOCITY_THRESHOLD)

print(f"\n{'='*80}")
print("ANALYSIS RESULTS")
print(f"{'='*80}")

# Overall statistics
total_runs = len(trajectories)
avg_runs_per_match = total_runs / len(MATCH_IDS)
total_space = trajectories['total_space_created'].sum()
avg_space_per_run = trajectories['total_space_created'].mean()
pitch_area = 105 * 68  # 7,140 m²
pct_of_pitch = (avg_space_per_run / pitch_area) * 100
avg_velocity = trajectories['max_velocity'].mean()
avg_velocity_kmh = avg_velocity * 3.6
avg_duration_seconds = trajectories['duration_frames'].mean() / 10

print(f"\nTotal runs detected: {total_runs}")
print(f"Average runs per match: {avg_runs_per_match:.1f}")
print(f"Total space created: {total_space:,.0f} m²")
print(f"Average space per run: {avg_space_per_run:,.0f} m² ({pct_of_pitch:.1f}% of pitch)")
print(f"Average velocity: {avg_velocity:.2f} m/s ({avg_velocity_kmh:.1f} km/h)")
print(f"Average duration: {avg_duration_seconds:.1f} seconds")


=== ANALYZING ALL MATCHES (threshold: 5.0 m/s) ===



Processing matches:   0%|          | 0/10 [00:00<?, ?it/s]

✓ Detected 12,920 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 20 unique players

Analyzing 12,920 run frames (team possession filter)...


Analyzing runs: 100%|██████████| 12920/12920 [00:08<00:00, 1610.27it/s]


✓ Analyzed 199 off-ball runs (space for ball carrier only)
✓ Detected 12,068 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 30 unique players

Analyzing 12,068 run frames (team possession filter)...


Processing matches:  10%|█         | 1/10 [00:19<02:57, 19.73s/it]

✓ Analyzed 362 off-ball runs (space for ball carrier only)
✓ Detected 11,520 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 20 unique players

Analyzing 11,520 run frames (team possession filter)...


Analyzing runs: 100%|██████████| 11520/11520 [00:07<00:00, 1624.29it/s]


✓ Analyzed 293 off-ball runs (space for ball carrier only)
✓ Detected 12,030 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 28 unique players

Analyzing 12,030 run frames (team possession filter)...


Processing matches:  20%|██        | 2/10 [00:37<02:29, 18.63s/it]

✓ Analyzed 129 off-ball runs (space for ball carrier only)
✓ Detected 13,048 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 20 unique players

Analyzing 13,048 run frames (team possession filter)...


Analyzing runs: 100%|██████████| 13048/13048 [00:09<00:00, 1420.36it/s]


✓ Analyzed 308 off-ball runs (space for ball carrier only)
✓ Detected 12,393 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 27 unique players

Analyzing 12,393 run frames (team possession filter)...


Processing matches:  30%|███       | 3/10 [00:57<02:15, 19.37s/it]

✓ Analyzed 250 off-ball runs (space for ball carrier only)
✓ Detected 10,335 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 21 unique players

Analyzing 10,335 run frames (team possession filter)...


Analyzing runs: 100%|██████████| 10335/10335 [00:04<00:00, 2266.11it/s]


✓ Analyzed 97 off-ball runs (space for ball carrier only)
✓ Detected 12,230 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 26 unique players

Analyzing 12,230 run frames (team possession filter)...


Processing matches:  40%|████      | 4/10 [01:13<01:48, 18.05s/it]

✓ Analyzed 310 off-ball runs (space for ball carrier only)
✓ Detected 9,544 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 21 unique players

Analyzing 9,544 run frames (team possession filter)...


Analyzing runs: 100%|██████████| 9544/9544 [00:05<00:00, 1682.98it/s]


✓ Analyzed 256 off-ball runs (space for ball carrier only)
✓ Detected 6,680 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 28 unique players

Analyzing 6,680 run frames (team possession filter)...


Processing matches:  50%|█████     | 5/10 [01:29<01:26, 17.23s/it]

✓ Analyzed 195 off-ball runs (space for ball carrier only)
✓ Detected 8,340 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 22 unique players

Analyzing 8,340 run frames (team possession filter)...


Analyzing runs: 100%|██████████| 8340/8340 [00:04<00:00, 2079.10it/s]


✓ Analyzed 63 off-ball runs (space for ball carrier only)
✓ Detected 8,292 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 26 unique players

Analyzing 8,292 run frames (team possession filter)...


Processing matches:  60%|██████    | 6/10 [01:42<01:02, 15.74s/it]

✓ Analyzed 176 off-ball runs (space for ball carrier only)
✓ Detected 8,364 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 20 unique players

Analyzing 8,364 run frames (team possession filter)...


Analyzing runs: 100%|██████████| 8364/8364 [00:04<00:00, 1765.96it/s]


✓ Analyzed 211 off-ball runs (space for ball carrier only)
✓ Detected 9,318 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 29 unique players

Analyzing 9,318 run frames (team possession filter)...


Processing matches:  70%|███████   | 7/10 [01:57<00:46, 15.49s/it]

✓ Analyzed 335 off-ball runs (space for ball carrier only)
✓ Detected 13,459 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 20 unique players

Analyzing 13,459 run frames (team possession filter)...


Analyzing runs: 100%|██████████| 13459/13459 [00:07<00:00, 1712.09it/s]


✓ Analyzed 401 off-ball runs (space for ball carrier only)
✓ Detected 13,513 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 30 unique players

Analyzing 13,513 run frames (team possession filter)...


Processing matches:  80%|████████  | 8/10 [02:17<00:33, 16.92s/it]

✓ Analyzed 265 off-ball runs (space for ball carrier only)
✓ Detected 12,759 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 20 unique players

Analyzing 12,759 run frames (team possession filter)...


Analyzing runs: 100%|██████████| 12759/12759 [00:06<00:00, 1864.34it/s]


✓ Analyzed 205 off-ball runs (space for ball carrier only)
✓ Detected 14,530 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 28 unique players

Analyzing 14,530 run frames (team possession filter)...


Processing matches:  90%|█████████ | 9/10 [02:37<00:17, 17.75s/it]

✓ Analyzed 340 off-ball runs (space for ball carrier only)
✓ Detected 13,953 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 20 unique players

Analyzing 13,953 run frames (team possession filter)...


Analyzing runs: 100%|██████████| 13953/13953 [00:06<00:00, 2301.18it/s]


✓ Analyzed 220 off-ball runs (space for ball carrier only)
✓ Detected 15,069 frames with velocity >= 5.0 m/s and duration >= 3.0s
✓ From 25 unique players

Analyzing 15,069 run frames (team possession filter)...


Processing matches: 100%|██████████| 10/10 [02:56<00:00, 17.68s/it]


✓ Analyzed 336 off-ball runs (space for ball carrier only)

✓ Total runs: 523
✓ Avg velocity: 6.49 m/s
✓ Avg space created: 5692 m²

ANALYSIS RESULTS

Total runs detected: 523
Average runs per match: 52.3
Total space created: 2,976,900 m²
Average space per run: 5,692 m² (79.7% of pitch)
Average velocity: 6.49 m/s (23.3 km/h)
Average duration: 0.9 seconds


## Results

### Overall Findings
Across 10 matches, **523 distinct off-ball runs** were detected (average 52.3 runs/match), creating an average of **5,692 m²** of space per run for ball carriers.

**Key Metrics:**
- Average space creation: **5,692 m²** per run
- Average run velocity: **6.49 m/s** (23.3 km/h)
- Average run duration: **0.9 seconds**

In [7]:
# Build player name dictionary
all_player_names = {}
for match_id in trajectories['match_id'].unique():
    data = load_match_data(match_id)
    for p in data['metadata']['players']:
        player_id = p['id']
        first = p.get('first_name', '')
        last = p.get('last_name', '')
        team_id = p['team_id']
        home_id = data['metadata']['home_team']['id']
        team_name = (data['metadata']['home_team']['name'] if team_id == home_id 
                    else data['metadata']['away_team']['name'])
        all_player_names[player_id] = {'name': f"{first} {last}".strip(), 'team': team_name}

# Player statistics
player_stats = trajectories.groupby('player_id').agg({
    'total_space_created': ['count', 'sum', 'mean'],
    'match_id': 'nunique'
}).reset_index()
player_stats.columns = ['player_id', 'num_runs', 'total_space', 'avg_space_per_run', 'matches']
player_stats['runs_per_match'] = player_stats['num_runs'] / player_stats['matches']
player_stats['space_per_match'] = player_stats['total_space'] / player_stats['matches']

print(f"\n{'='*80}")
print("TOP PERFORMERS")
print(f"{'='*80}")

# Top 3 categories
top_runs = player_stats.nlargest(1, 'runs_per_match').iloc[0]
pid = int(top_runs['player_id'])
pinfo = all_player_names.get(pid, {'name': f'Player {pid}', 'team': 'Unknown'})
print(f"\n1. Most Runs/Match: {pinfo['name']} ({pinfo['team']})")
print(f"   {top_runs['runs_per_match']:.1f} runs/match ({int(top_runs['matches'])} matches) | {top_runs['space_per_match']:,.0f} m²/match")

top_space = player_stats.nlargest(1, 'space_per_match').iloc[0]
pid = int(top_space['player_id'])
pinfo = all_player_names.get(pid, {'name': f'Player {pid}', 'team': 'Unknown'})
print(f"\n2. Most Space/Match: {pinfo['name']} ({pinfo['team']})")
print(f"   {top_space['space_per_match']:,.0f} m²/match ({int(top_space['matches'])} matches) | {top_space['runs_per_match']:.1f} runs/match")

player_stats_eff = player_stats[player_stats['num_runs'] >= 5]
if len(player_stats_eff) > 0:
    top_eff = player_stats_eff.nlargest(1, 'avg_space_per_run').iloc[0]
    pid = int(top_eff['player_id'])
    pinfo = all_player_names.get(pid, {'name': f'Player {pid}', 'team': 'Unknown'})
    print(f"\n3. Most Efficient (min 5 runs): {pinfo['name']} ({pinfo['team']})")
    print(f"   {top_eff['avg_space_per_run']:,.0f} m²/run ({int(top_eff['matches'])} matches) | {top_eff['runs_per_match']:.1f} runs/match")


TOP PERFORMERS

1. Most Runs/Match: Thomas Aquilina (Newcastle United Jets FC)
   8.0 runs/match (1 matches) | 22,462 m²/match

2. Most Space/Match: Kai Trewin (Melbourne City FC)
   94,171 m²/match (1 matches) | 2.0 runs/match

3. Most Efficient (min 5 runs): Jordi Valadon (Melbourne Victory Football Club)
   25,317 m²/run (2 matches) | 2.5 runs/match


## Visualizations

### Figure 1: Voronoi Tessellation Methodology
![Voronoi Example](figs/voronoi_example.png)

Voronoi tessellation showing controlled space per player. The ball carrier (gold square) gains exploitable space when teammates make off-ball runs, pulling defenders away.

### Figure 2: Run Trajectory Patterns
![Trajectory Visualization](figs/trajectory_visualization.png)

All off-ball runs from Melbourne Victory showing normalized trajectories. Green markers indicate run start, red markers show run end. Arrows reveal coordinated movement patterns creating space in attacking third.

## Conclusion

This transparent, Voronoi-based approach successfully quantifies off-ball run effectiveness using broadcast tracking data. The methodology reveals measurable spatial advantages created by coordinated off-ball movement, with runs creating average exploitable space equivalent to 80% of pitch area for ball carriers.

### Limitations
Broadcast tracking captures only 51% of on-pitch moments due to camera coverage constraints. Future work should incorporate GPS data for complete coverage and extend analysis to pressing scenarios (runs during opponent possession).

### Impact
This interpretable framework enables coaches to evaluate off-ball movement patterns quantitatively, informing tactical preparation and player development without requiring complex machine learning infrastructure.

### Key Contributions
- **Transparency-first methodology:** No black-box ML, fully interpretable metrics
- **Actionable insights:** Quantifies previously unmeasured tactical elements
- **Scalable framework:** Applicable to any broadcast tracking data